File: Hazelcast.Net/Hazelcast.Client.Protocol.Codec/BuiltIn/WanReplicationRefCodec.java
/*
 * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hazelcast.client.impl.protocol.codec.builtin;

import com.hazelcast.client.impl.protocol.ClientMessage;
import com.hazelcast.config.WanReplicationRef;
import com.hazelcast.nio.Bits;

import java.util.List;
import java.util.ListIterator;

import static com.hazelcast.client.impl.protocol.ClientMessage.BeginFrame;
import static com.hazelcast.client.impl.protocol.ClientMessage.EndFrame;
import static com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.fastForwardToEndFrame;
import static com.hazelcast.client.impl.protocol.codec.builtin.FixedSizeTypesCodec.DecodeBoolean;
import static com.hazelcast.client.impl.protocol.codec.builtin.FixedSizeTypesCodec.EncodeBoolean;

public final class WanReplicationRefCodec {
    private const int REPUBLISHING_ENABLED_OFFSET = 0;
    private const int InitialFrameSize = REPUBLISHING_ENABLED_OFFSET + Bits.BOOLEAN_SIZE_IN_BYTES;

    private WanReplicationRefCodec() {
    }

    public static void Encode(ClientMessage clientMessage, WanReplicationRef wanReplicationRef) {
        clientMessage.add(BeginFrame);

        ClientMessage.Frame initialFrame = new ClientMessage.Frame(new byte[InitialFrameSize]);
        EncodeBoolean(initialFrame.content, REPUBLISHING_ENABLED_OFFSET, wanReplicationRef.isRepublishingEnabled());
        clientMessage.add(initialFrame);

        StringCodec.Encode(clientMessage, wanReplicationRef.getName());
        StringCodec.Encode(clientMessage, wanReplicationRef.getMergePolicy());
        ListMultiFrameCodec.EncodeNullable(clientMessage, wanReplicationRef.getFilters(), StringCodec::Encode);

        clientMessage.add(EndFrame);
    }

    public static WanReplicationRef Decode(ref ClientMessage.FrameIterator iterator) {
        // begin frame
        iterator.next();

        ClientMessage.Frame initialFrame = iterator.next();
        boolean republishingEnabled = DecodeBoolean(initialFrame.content, REPUBLISHING_ENABLED_OFFSET);

        String name = StringCodec.Decode(iterator);
        String mergePolicy = StringCodec.Decode(iterator);
        List<String> filters = ListMultiFrameCodec.DecodeNullable(iterator, StringCodec::Decode);

        fastForwardToEndFrame(iterator);

        return new WanReplicationRef(name, mergePolicy, filters, republishingEnabled);
    }
}

File: Hazelcast.Net/Hazelcast.Client.Protocol.Codec/custom/WanReplicationRefCodec.cs
/*
 * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections;
using System.Collections.Generic;
using Hazelcast.Client.Protocol;
using Hazelcast.Client.Protocol.Codec.BuiltIn;
using Hazelcast.Client.Protocol.Util;
using Hazelcast.IO;
using static Hazelcast.Client.Protocol.Codec.BuiltIn.FixedSizeTypesCodec;
using static Hazelcast.Client.Protocol.ClientMessage;
using static Hazelcast.IO.Bits;

namespace Hazelcast.Client.Protocol.Codec.Custom
{
    /*
    * This file is auto-generated by the Hazelcast Client Protocol Code Generator.
    * To change this file, edit the templates or the protocol
    * definitions on the https://github.com/hazelcast/hazelcast-client-protocol
    * and regenerate it.
    */
    internal static class WanReplicationRefCodec 
    {
        private const int RepublishingEnabledFieldOffset = 0;
        private const int InitialFrameSize = RepublishingEnabledFieldOffset + BooleanSizeInBytes;

        public static void Encode(ClientMessage clientMessage, com.hazelcast.config.WanReplicationRef wanReplicationRef) 
        {
            clientMessage.Add(BeginFrame);

            var initialFrame = new Frame(new byte[InitialFrameSize]);
            EncodeBool(initialFrame.Content, RepublishingEnabledFieldOffset, wanReplicationRef.IsRepublishingEnabled);
            clientMessage.Add(initialFrame);

            StringCodec.Encode(clientMessage, wanReplicationRef.Name);
            StringCodec.Encode(clientMessage, wanReplicationRef.MergePolicy);
            ListMultiFrameCodec.EncodeNullable(clientMessage, wanReplicationRef.Filters, StringCodec.Encode);

            clientMessage.Add(EndFrame);
        }

        public static com.hazelcast.config.WanReplicationRef Decode(ref FrameIterator iterator) 
        {
            // begin frame
            iterator.Next();

            ref var initialFrame = ref iterator.Next();
            var republishingEnabled = DecodeBool(initialFrame.Content, RepublishingEnabledFieldOffset);

            var name = StringCodec.Decode(ref iterator);
            var mergePolicy = StringCodec.Decode(ref iterator);
            var filters = ListMultiFrameCodec.DecodeNullable(ref iterator, StringCodec.Decode);

            CodecUtil.FastForwardToEndFrame(ref iterator);

            return new com.hazelcast.config.WanReplicationRef(name, mergePolicy, filters, republishingEnabled);
        }
    }
}